src/main/scala/analysis/RegionInjector.scala

```diff
@@ -145,16 +163,148 @@ class RegionInjector(program: Program, mmm: MemoryModelMap) {
+    queue.enqueue(program.mainProcedure)
+
+    while (queue.nonEmpty) {
+      val procedure = queue.dequeue()
+      for {
+        b <- procedure.blocks
+        s <- b.statements
+      } {
+        s match {
+          case store: MemoryStore =>
+            val dsg = DSATopDown(procedure)
+            val cell = dsg.find(dsg.accessIndexToSlice(store)).cell
+            val mergedRegion = if (mergedRegions.contains(cell)) {
+              mergedRegions(cell)
+            } else {
+              val region = createRegion(cell)
+              mergedRegions(cell) = region
+              region
+            }
+            val newMemory = replaceMemory(store.mem, mergedRegion)
+            store.mem = newMemory
+          case load: MemoryLoad =>
+            val dsg = DSATopDown(procedure)
+            val cell = dsg.find(dsg.accessIndexToSlice(load)).cell
+            val mergedRegion = if (mergedRegions.contains(cell)) {
+              mergedRegions(cell)
+            } else {
+              val region = createRegion(cell)
+              mergedRegions(cell) = region
+              region
+            }
+            val newMemory = replaceMemory(load.mem, mergedRegion)
+            load.mem = newMemory
+          case _ =>
+        }
+      }
+      visited.add(procedure)
+      for (call <- procedure.calls) {
+        if (!queue.contains(call) && !visited.contains(call)) {
+          queue.enqueue(call)
+        }
+      }
+    }
+    transformMemorySections(program)
+  }
+
+  private def createRegion(cell: Cell): MergedRegionDSA = {
+    val stack = cell.node.get.flags.stack
+    val name = if (stack) {
+      stackCounter += 1
+      s"stack$$${stackCounter}"
+    } else {
+      sharedMemoryCounter += 1
+      s"mem$$${sharedMemoryCounter}"
+    }
+    MergedRegionDSA(name, cell, stack)
+  }
+
+  private def replaceMemory(memory: Memory, mergedRegion: MergedRegionDSA): Memory = {
+    if (mergedRegion.stack) {
+      StackMemory(mergedRegion.name, memory.addressSize, memory.valueSize)
+    } else {
+      SharedMemory(mergedRegion.name, memory.addressSize, memory.valueSize)
+    }
+  }
+
+  private def transformMemorySections(program: Program): Unit = {
+    // if any addressRange entries point to a mergedRegion: grab the bytes
+
+    val dsg = DSATopDown(program.mainProcedure)
+
+    // need to check if nodes match too
+
+    dsg.globalMapping.foreach { (range, field) =>
+      val cell = dsg.find(field.node.cells(0))
+      if (mergedRegions.contains(cell)) {
+        program.initialMemoryLookup(range.start) match {
+          case Some(section) =>
+            val size = (range.end - range.start).toInt
+            val bytes = section.getBytes(range.start, size)
+            // should probably check that region is entirely contained within section but shouldn't happen in practice?
+            val newSection = MemorySection(mergedRegions(cell).name, range.start, size, bytes, section.readOnly, Some(mergedRegions(cell)))
```

Comment:
each global may consistent of multiple disjoint cells each representing different regions. The size of memory section should be the largest accessed size of the cell?

Reply:
When is the `size` of a `Cell` different to the `largestAccessedSize`?

I did realise that my approach doesn't currently take into account the case where an access should have two cells associated with it (due to overlapping accesses) but it was not really possible to figure that out yet since that hasn't yet been implemented on this branch..